File: leetcode/problems/cpp/daily-temperatures.cpp
// by @codeAbinash
// Time : O(n)
// Space : O(n)

#include "vector"
using namespace std;

class Solution {
  public:
   vector<int> dailyTemperatures(vector<int>& temperatures) {
      int size = temperatures.size();
      vector<int> ans(size, 0);
      vector<int> stack;
      for (int i = 0; i < size; i++) {
         while (!stack.empty() && temperatures[stack.back()] < temperatures[i]) {
            ans[stack.back()] = i - stack.back();
            stack.pop_back();
         }
         stack.push_back(i);
      }
      return ans;
   }
};

File: leetcode/problems/rs/daily-temperatures.rs
// by @codeAbinash
// Time : O(n)
// Space : O(n)

pub struct Solution;

impl Solution {
   pub fn daily_temperatures(temperatures: Vec<i32>) -> Vec<i32> {
      let mut stack = Vec::new();
      let mut result = vec![0; temperatures.len()];
      for i in 0..temperatures.len() {
         while !stack.is_empty() && temperatures[i] > temperatures[*stack.last().unwrap()] {
            let last = stack.pop().unwrap();
            result[last] = (i - last) as i32;
         }
         stack.push(i);
      }
      result
   }
}
